_posts/2022-08-17-using-the-kani-rust-verifier-on-tokio-bytes.md

```diff
@@ -0,0 +1,556 @@
+3. Use [`mem::forget`](https://doc.rust-lang.org/std/mem/fn.forget.html) to ensure that `vec`'s destructor will *not* run when `from_vec` returns (notice that `vec` is moved into `from_vec`).
+This is essential to ensure that the buffer pointed to by `ptr` remains valid.
+4. Set `data` to `KIND_VEC` (`=0b1`), which is subsequently cast to a raw pointer `*mut Shared`
+
+This is interesting, but unfortunately there is not a great deal more we can say for `is_valid_kind_vec()`.
+If we have an instance using `KIND_VEC` then we know that the `ptr` and `cap` must be related to the "original vector".
+For example, immediately after construction, we know `ptr` must be exactly `vec`'s buffer and similarly for `cap`.
+However, we no longer have access to this `vec`, so for now we will make our well-formedness check a no-op.
+[Don't worry, we will return to this later!]
+
+```rust
+fn is_valid_kind_vec(&self) -> bool {
+    assert!(self.kind() == KIND_VEC);
+    true
+}
+```
+
+Let's now turn to the `KIND_ARC` case.
+Here we are more fortunate because we have the `Shared` object pointed to by the `data` field of the `BytesMut` instance.
+As a reminder, here's what the `Shared` struct looks like:
+
+```rust
+struct Shared {
+    vec: Vec<u8>,
+    original_capacity_repr: usize,
+    ref_count: AtomicUsize,
+}
+```
+
+The `ref_count` is the number of references (including the instance, itself) to the backing buffer.
+For this reason, we know that `ref_count` *must* be at least `1`.
+Secondly, `vec` is the "original vector" that we have derived our instance's slice from.
+[We will ignore `original_capacity_repr` which is used to ensure that `reserve` always allocates sufficient bytes.]
+Having this "in our hands" is useful because it must be the case that the `ptr` is within `vec`'s buffer, and similarly for capacity.
+Putting this together:
+
+```rust
+fn is_valid_kind_arc(&self) -> bool {
+    assert!(self.kind() == KIND_ARC);
+    let shared: *mut Shared = self.data as _;
+    unsafe {
+        let ref_count = (*shared).ref_count.load(Ordering::Relaxed);
+        let valid_ref_count = 1 <= ref_count;
+
+        let vec_ptr = (*shared).vec.as_mut_ptr();
+        let vec_cap = (*shared).vec.capacity();
+        let vec_end_ptr = vec_ptr.offset(vec_cap as isize);
+        let ptr = self.ptr.as_ptr();
+        let ptr_in_bounds = vec_ptr <= ptr && ptr <= vec_end_ptr;
+        let cap_in_bounds = self.cap <= vec_cap;
+
+        valid_ref_count && ptr_in_bounds && cap_in_bounds
+    }
+}
+```
+
+This gives us an idea!
+What if we could keep track of the original vector for `KIND_VEC` too?
+The use of program state that is just for verification is known as *ghost state*, originally called [auxiliary variables](https://dl.acm.org/doi/pdf/10.1145/360051.360224).
```

Comment:
Should we note that this will change the layout in Kani

Reply:
Added ", although it will change the layout in Kani" after "hence has not runtime cost" below, but I'm not sure if that'll be confusing. What do you think?

Reply:
I'd maybe put that as a footnote?